File: src/main/java/org/albaross/agents4j/extraction/bases/ModularKnowledgeBase.java
package org.albaross.agents4j.extraction.bases;

import lombok.RequiredArgsConstructor;
import org.albaross.agents4j.extraction.KnowledgeBase;
import org.albaross.agents4j.extraction.data.Rule;

import java.util.*;

import static java.util.Collections.emptyList;
import static java.util.Comparator.reverseOrder;
import static java.util.stream.Collectors.toList;

public class ModularKnowledgeBase<A> implements KnowledgeBase<A> {

    private final TreeMap<Long, TreeMap<Double, TreeSet<Rule<A>>>> base = new TreeMap<>(reverseOrder());
    private int size = 0;

    @Override
    public int size() {
        return size;
    }

    @Override
    public Collection<Rule<A>> reason(Set<String> state) {
        if (base.isEmpty())
            return emptyList();

        for (TreeMap<Double, TreeSet<Rule<A>>> level : base.values()) {
            for (TreeSet<Rule<A>> module : level.values()) {

                List<Rule<A>> active = module.stream()
                        .filter(r -> state.containsAll(r.getPremise()))
                        .collect(toList());

                if (!active.isEmpty())
                    return active;
            }
        }

        return emptyList();
    }

    @Override
    public void add(Rule<A> rule) {
        long dim = rule.getPremise().size();
        double conf = rule.getConfidence();

        TreeMap<Double, TreeSet<Rule<A>>> level = base.get(dim);
        if (level == null) base.put(dim, level = new TreeMap<>(reverseOrder()));

        TreeSet<Rule<A>> module = level.get(conf);
        if (module == null) level.put(conf, module = new TreeSet<>());

        if (module.add(rule)) size++;
    }

    @Override
    public void remove(Rule<A> rule) {
        long dim = rule.getPremise().size();
        double conf = rule.getConfidence();

        TreeMap<Double, TreeSet<Rule<A>>> level = base.get(dim);
        if (level == null) return;

        Set<Rule<A>> module = level.get(conf);
        if (module == null) return;

        if (module.remove(rule)) size--;

        if (module.isEmpty())
            level.remove(conf);

        if (level.isEmpty())
            base.remove(dim);
    }

    @Override
    public void clear() {
        base.clear();
    }

    @Override
    public Iterator<Collection<Rule<A>>> iterator() {
        return new ModularIterator<>(base.descendingMap().values().iterator());
    }

    @RequiredArgsConstructor
    private static class ModularIterator<A> implements Iterator<Collection<Rule<A>>> {

        private final Iterator<TreeMap<Double, TreeSet<Rule<A>>>> iterator;

        @Override
        public boolean hasNext() {
            return iterator.hasNext();
        }

        @Override
        public Collection<Rule<A>> next() {
            return iterator.next().values().stream()
                    .flatMap(Collection::stream)
                    .collect(toList());
        }
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder();
        sb.append("--------------------\n");

        base.descendingMap().forEach((dim, level) -> {
            level.forEach((conf, module) -> module.forEach(r -> r.appendTo(sb).append("\n")));
            sb.append("--------------------\n");
        });

        return sb.toString();
    }

}

File: src/main/kotlin/bases/ModularKnowledgeBase.kt
package org.albaross.agents4j.extraction.bases

import org.albaross.agents4j.extraction.KnowledgeBase
import org.albaross.agents4j.extraction.data.Rule
import java.util.*

class ModularKnowledgeBase<A> : KnowledgeBase<A> {

    private val base = TreeMap<Int, TreeMap<Double, TreeSet<Rule<A>>>>()
    private var count = 0L

    override val ruleCount: Long
        get() = count

    override val levelCount: Int
        get() = base.keys.firstOrNull() ?: 0

    override fun reason(state: Set<String>): Collection<Rule<A>> {
        for ((_, level) in base) {
            for ((_, module) in level) {
                // find all matching rules in this module
                val active = module.filter { state.containsAll(it.state) }

                // return in case any rules are matching
                if (!active.isEmpty())
                    return active
            }
        }

        return emptyList()
    }

    override fun add(rule: Rule<A>) {
        val dim = rule.state.size
        val conf = rule.weight

        // find matching module
        val level = base.getOrPut(dim) { TreeMap(reverseOrder()) }
        val module = level.getOrPut(conf) { TreeSet() }

        // add rule to module
        if (module.add(rule)) count++
    }

    override fun remove(rule: Rule<A>) {
        val dim = rule.state.size
        val conf = rule.weight

        // find matching module
        val level = base[dim] ?: return
        val module = level[conf] ?: return

        // remove rule from module
        if (module.remove(rule)) count--

        // clean up empty modules and levels
        if (module.isEmpty()) level.remove(conf)
        if (level.isEmpty()) base.remove(dim)
    }

    override fun iterator(): Iterator<Collection<Rule<A>>> = ModularIterator(base.values.iterator())

    override fun clear() {
        base.clear()
        count = 0L
    }

    override fun toString() = base.toString()

}

private class ModularIterator<A>(private val backing: Iterator<Map<*, Collection<Rule<A>>>>) : Iterator<Collection<Rule<A>>> {

    override fun hasNext() = backing.hasNext()

    override fun next() = backing.next().values.flatMap { it.toList() }

}